src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/scheduler/BlobProcessor.java

```diff
@@ -0,0 +1,143 @@
+ * 4. Save PDF files in document storage.
+ * 5. Update status and doc urls in DB.
+ */
+@Component
+@ConditionalOnProperty(value = "scheduling.enabled", matchIfMissing = true)
+public class BlobProcessor {
+
+    private static final Logger log = LoggerFactory.getLogger(BlobProcessor.class);
+
+    private final CloudBlobClient cloudBlobClient;
+    private final EnvelopeRepository envelopeRepository;
+    private final DocumentManagementService documentManagementService;
+
+    @Autowired
+    public BlobProcessor(
+        CloudBlobClient cloudBlobClient,
+        EnvelopeRepository envelopeRepository,
+        DocumentManagementService documentManagementService
+    ) {
+        this.cloudBlobClient = cloudBlobClient;
+        this.envelopeRepository = envelopeRepository;
+        this.documentManagementService = documentManagementService;
+    }
+
+    @Scheduled(fixedDelayString = "${scan.delay}")
+    public void processBlobs() {
+        cloudBlobClient.listContainers()
+            .forEach(cloudBlobContainer ->
+                processBlobsPerContainer(cloudBlobContainer.getName())
+            );
+    }
+
+    private void processBlobsPerContainer(String containerName) {
+        log.info("Processing blobs for container {} ", containerName);
+
+        try {
+            processZipFiles(cloudBlobClient.getContainerReference(containerName));
+        } catch (URISyntaxException | StorageException e) {
+            log.error("Exception occurred while retrieving container reference for " + containerName, e);
+            // TODO Raise an event in DB or track in appinsights if this happens.
+        }
+    }
+
+    private void processZipFiles(CloudBlobContainer container) {
+        for (ListBlobItem blobItem : container.listBlobs()) {
+            List<Pdf> pdfFiles = new ArrayList<>();
+            String zipFilename = FilenameUtils.getName(blobItem.getUri().toString());
+            byte[] metadataStream = null;
+
+            try {
+                BlobInputStream blobInputStream =
+                    container.getBlockBlobReference(zipFilename).openInputStream();
+
+                //Zip file will include metadata.json and collection of pdf documents
+                try (ZipInputStream zis = new ZipInputStream(blobInputStream)) {
+                    ZipEntry zipEntry;
+                    while ((zipEntry = zis.getNextEntry()) != null) {
+                        switch (FilenameUtils.getExtension(zipEntry.getName())) {
+                            case "json":
+                                metadataStream = toByteArray(zis);
+                                break;
+                            case "pdf":
+                                Pdf pdf = new Pdf(zipEntry.getName(), toByteArray(zis));
+                                pdfFiles.add(pdf);
+                                break;
+                            default:
+                                throw new NonPdfFileFoundException("Zip file contains non pdf documents");
```

Comment:
Giving it is inside while loop does it really make sense to throw an exception here? You can have not only json/pdf files in zip and this would just break the process. Perhaps better log a warning and validate number of pdfs extracted against parsed metafile (line 111-112)

Reply:
`UnsupportedDataException`?